scala/lambdas/ingest-validate-generic-ingest-inputs/src/main/scala/uk/gov/nationalarchives/ingestvalidategenericingestinputs/Lambda.scala

```diff
@@ -0,0 +1,261 @@
+        lazy val keyOfResultsFile = s"$keyWithoutOriginalFileName/metadata-entries-with-errors.json"
+        lazy val fullLocationOfResultsFile = s"${bucket}/$keyOfResultsFile"
+        val jsonStructureErrorMessage =
+          if numOfJsonStructureErrors > 0 then
+            val s = if numOfJsonStructureErrors == 1 then "" else "s"
+            s"$numOfJsonStructureErrors thing$s wrong with the structure of the metadata.json for batchId '${input.batchId}'; the results can be found here: $fullLocationOfResultsFile\n\n"
+          else ""
+
+        val entriesErrorMessage =
+          if numOfEntriesWithFailures > 0 then
+            val (entryOrEntries, hasOrHave) = if numOfEntriesWithFailures == 1 then ("entry", "has") else ("entries", "have")
+            s"$numOfEntriesWithFailures $entryOrEntries (objects) in the metadata.json for batchId '${input.batchId}' $hasOrHave failed validation; the results can be found here: $fullLocationOfResultsFile"
+          else ""
+
+        for {
+          _ <- if jsonStructureErrorMessage.isEmpty then IO.unit else log(jsonStructureErrorMessage)
+          _ <-
+            if entriesErrorMessage.isEmpty then IO.unit
+            else
+              log(entriesErrorMessage) >> entriesThatFailedValidation.parTraverse { entry =>
+                val entryId = entry("id") match {
+                  case ValidValidationResult(id) => id
+                  case InvalidValidationResult(listOfErrors) =>
+                    val potentialIdValues = listOfErrors.collect { case valueError: ValueError => valueError.valueThatCausedError }
+                    potentialIdValues.headOption.getOrElse("idUnknown")
+                }
+
+                val fieldNamesWithErrors = entry.collect { case (name, value) if value.getClass.getSimpleName == "InvalidValidationResult" => name }.mkString(", ")
+                val mapToSendAsJson = Map("id" -> entryId, "fieldNamesWithErrors" -> fieldNamesWithErrors, "locationOfFile" -> fullLocationOfResultsFile)
+
+                log(mapToSendAsJson.asJson.noSpaces)
+              }
+          allJsonErrors = entriesThatFailedValidation.+:(jsonStructuralErrors)
+          _ <- Stream
+            .eval(IO.pure(allJsonErrors.asJson.noSpaces))
+            .map(s => ByteBuffer.wrap(s.getBytes()))
+            .toPublisherResource
+            .use { pub =>
+              s3Client.upload(bucket, keyOfResultsFile, FlowAdapters.toPublisher(pub))
+            }
+            .map(_ => ())
+
+          _ <- IO.raiseError(JsonValidationException(jsonStructureErrorMessage + entriesErrorMessage))
+        } yield ()
+      }
+
+    } yield StateOutput(input.batchId, input.metadataPackage)
+
+  private def validateAgainstSchema(metadataJson: String) =
+    for {
+      metadataJsonAsUjson <- IO(read(metadataJson).arr.toList)
+      entriesGroupedByType = metadataJsonAsUjson.foldLeft(Map[String, List[Value]]()) { (typesGrouped, entry) =>
```

Comment:
Ah, ok. I think, I used try because I thought, "what if `type` was missing or it was another type other than `String` (by some sort of mistake)?" So, I'll use IO error handling for this

Reply:
But the code in the comment will give you "UnknownType" if the type is missing or if it's another type because the first case will fail.  You're not throwing any errors or anything if the type is missing or unknown so you shouldn't need any error handling should you?

Reply:
Ah, no, sorry, my comment should have been clearer; I wasn't refuting what you said, I was just mentioning the fact that I used `Try` because I was doing `entry(typeField).str`, it didn't cross my mind that you could do a `value.obj.contains` and now that you mentioned `recoverWith` in the other comment you made, now I know that I don't need to use `Try` anymore. 👍